modules/ROOT/pages/cross_cutting/i18n.adoc

```diff
@@ -17,31 +19,155 @@ Also consider:
+When a messageKey should be translated into a locale message, the translator looks up the link:https://docs.oracle.com/javase/7/docs/api/java/util/ResourceBundle.html[`ResourceBundle`] for the locale and returns the message from it.
+Then link:https://unicode-org.github.io/icu/userguide/icu4j/[ICU4J] is used to format the message and replace variables. In this example, maps are used to allow named parameters. When using the native Message Formatter, this is not possible. 
+[source, java]
+----
+public class MessageTranslator {
+    private final static ConcurrentHashMap<String, MessageTranslator> instances = new ConcurrentHashMap <>(); // <1>
+    private final ConcurrentHashMap<Locale, ResourceBundle> bundleMap = new ConcurrentHashMap<>(); // <2>
+    private final String BASE_DIRECTORY_NAME = "i18n";
+    private final String MESSAGE_PROPERTY_PREFIX = "message";
+    private final String messageBundleName; // <3>
+
+    private MessageTranslator(String messageBundleName){...}
+
+    public static MessageTranslator getInstance(String messageBundleName) {// <1>
+        return instances.computeIfAbsent(messageBundleName, MessageTranslator::new);
+    }
+
+    private String getPattern(String patternKey, Locale locale){// <4>
+        ResourceBundle bundle = bundleMap.computeIfAbsent(locale, (key) ->
+             ResourceBundle.getBundle(
+                String.join("/", BASE_DIRECTORY_NAME, messageBundleName, MESSAGE_PROPERTY_PREFIX),
+                key,
+                Thread.currentThread().getContextClassLoader()
+             )
+        );
+        return bundle.getString(patternKey);
+    }
+
+    public String translate(String patternKey, Locale locale, Map<String, Object> arguments){
+        String pattern = getPattern(patternKey, locale);
+        MessageFormat messageFormat = new MessageFormat(pattern, locale);
+        StringBuffer stringBuffer = new StringBuffer();
+        return messageFormat.format(arguments, stringBuffer, new FieldPosition(0)).toString();// <5>
+    }
+}
 ----
-// Don't use just the language (en)
-message_en.properties // LOCALE.EN
+<1> A `MessageTranslator` instance exists for each group of messages. Make sure to use thread safe code. 
+<2> A `MessageTranslator` instance has a link:https://docs.oracle.com/javase/7/docs/api/java/util/ResourceBundle.html[`ResourceBundle`] per locale.
+<3> The messageBundles are stored in the same folder with the pattern `i18n/<messageBundleName>/message_<LOCALE>.properties`
+<4> Returns the message from the locale-specific link:https://docs.oracle.com/javase/7/docs/api/java/util/ResourceBundle.html[`ResourceBundle`] using the bundleMap.
+<5> Uses the ICU's link:https://unicode-org.github.io/icu-docs/apidoc/dev/icu4j/com/ibm/icu/text/MessageFormat.html[`MessageFormat`] to format the message based on the arguments and the locale.
+--
+message_en_US.properties::
++
+--
+----
+welcome=Welcome {name}
+bookingDateShort=Your table is booked at {bookingDate, date, short}. <1>
```

Comment:
Sry, I hate being picky with that, but here it should be bookingDateLong and the pattern in the end need the same addition.